### PR TITLE
change(rpc): return an error from getblocktemplate method if Zebra is not synced to network tip

### LINKED ISSUE #5466: Return an error from getblocktemplate if Zebra isn't synced to the tip
## Motivation

`s-nomp` uses `getblocktemplate` to check if the node is synced to the tip:
https://github.com/s-nomp/node-stratum-pool/blob/d86ae73f8ff968d9355bb61aac05e0ebef36ccb5/lib/pool.js#L140

### Designs

- Use a `SyncStatus` to check that Zebra has finished syncing
- Allow the same clock skew as the Zcash network, which is 100 blocks, based on the standard rule:

> a full validator MUST NOT accept blocks with nTime more than two hours in the future according to its clock. This is not strictly a consensus rule because it is nondeterministic, and clock time varies between nodes.

https://zips.z.cash/protocol/protocol.pdf#blockheader

Completed work:
<details>

Get the chain tip and the estimated height, and see if they are close enough.
Here is some existing RPC code that gets those fields:
https://github.com/ZcashFoundation/zebra/blob/a2dba8cc5dd482531883ea982398cf0e36c6fcdd/zebra-rpc/src/methods.rs#L370-L397

</details>

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

We want to return an error if there's distance between Zebra's best chain tip height and the estimated network chain tip height because mining pools may use the `getblocktemplate` method to check if the node is synced.

Closes #5466

## Solution

- Return a server error if the estimated network height is higher than Zebra's chain tip height

## Review

This PR is part of regular scheduled work.

### Reviewer Checklist

  - [x] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [x] Are the PR labels correct?
  - [x] Does the code do what the ticket and PR says?
  - [x] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebra-chain/src/chain_tip.rs [hunk 1]
```diff
@@ -58,6 +58,21 @@ pub trait ChainTip {
...
         Some(estimator.estimate_height_at(now))
     }
+
+    /// Return an estimate of how many blocks there are ahead of zebra's best chain tip
+    /// until the network chain tip.
+    ///
+    /// The estimate is calculated based on the current local time, the block time of the best tip
+    /// and the height of the best tip.
+    ///
+    /// This estimate may be negative if the current local time is behind the chain tip block's timestamp.
```
Comment (dconnolly) [inline zebra-chain/src/chain_tip.rs:68]: Nice

Comment (arya2) [reply zebra-chain/src/chain_tip.rs:68]: Do we know why past times are handled this way?

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/chain_tip/network_chain_tip_height_estimator.rs#L122-L137

**Comment by dconnolly** [comment]
Restarting this, this failure may be related to flipping backend switches in gcloud related to OS Login:

https://github.com/ZcashFoundation/zebra/pull/5602

https://github.com/ZcashFoundation/zebra/actions/runs/3447560707/jobs/5754223166
image

**Comment by gustavovalverde** [comment]
> Restarting this, this failure may be related to flipping backend switches in gcloud related to OS Login

@dconnolly this is a missing sudo command. I added it in #5602 but any of you can add it here if needed (not sure why this didn't fail before)

**Review by dconnolly — DISMISSED** [review]
lgtm!

File: zebra-chain/src/chain_tip.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-chain/src/chain_tip.rs:68]: I don't recall, but I'd be willing to put money on "this is how zcashd handles it"

File: .github/workflows/deploy-gcp-tests.yml [hunk 2]
```diff
@@ -1266,6 +1266,18 @@ jobs:
...
+      # could cause the following error:
+      # `Login profile size exceeds 32 KiB. Delete profile values to make additional space`
+      - name: Delete temporal SSH keys
+        continue-on-error: true
+        run: |
+          for i in $(gcloud compute os-login ssh-keys list --format="table[no-heading](value.fingerprint)"); do 
+            echo "$i";
+            gcloud compute os-login ssh-keys remove --key "$i" || true;
+          done
+
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: @gustavovalverde should we do this in #5602 and merge?

Comment (arya2) [reply .github/workflows/deploy-gcp-tests.yml]: Seems to be needed for every PR to work so I reverted this for now and will wait for https://github.com/ZcashFoundation/zebra/pull/5602 to merge.

**Comment by arya2** [comment]
@Mergifyio update

**Comment by oxarbitrage** [comment]
This looks really good but i will lime to see a test with a chain synced up to tip and a call to `getblocktemplate` making sure we don't hit the error and the function is still usable.

We don't need to add this to CI but just a log here will be enough.

**Comment by arya2** [comment]
> This looks really good but i will lime to see a test with a chain synced up to tip and a call to `getblocktemplate` making sure we don't hit the error and the function is still usable.
> 
> We don't need to add this to CI but just a log here will be enough.

This helped catch an issue!

I'm thinking I want to add a test that syncs to the tip for getblocktemplate in CI too for testing https://github.com/ZcashFoundation/zebra/pull/5630 with the non-finalized state, but for now:

```json
{
  "jsonrpc": "2.0",
  "result": {
    "capabilities": [],
    "version": 4,
    "previousblockhash": "0000000000000000000000000000000000000000000000000000000000000000",
    "blockcommitmentshash": "0000000000000000000000000000000000000000000000000000000000000000",
    "lightclientroothash": "0000000000000000000000000000000000000000000000000000000000000000",
    "finalsaplingroothash": "0000000000000000000000000000000000000000000000000000000000000000",
    "defaultroots": {
      "merkleroot": "f00a46d38fccf8b48554c1f4651eec6bbc294c79ac2e2bc91e63bfcdc574eaa9",
      "chainhistoryroot": "0000000000000000000000000000000000000000000000000000000000000000",
      "authdataroot": "8231fdfe4e63846249d0880168deb69bcf7d81cc2f762937677f203fa8d8fd2a",
      "blockcommitmentshash": "0000000000000000000000000000000000000000000000000000000000000000"
    },
    "transactions": [],
    "coinbasetxn": {
      "data": "050000800a27a726b4d0d6c20000000082b31c00010000000000000000000000000000000000000000000000000000000000000000ffffffff040382b31c0000000004286bee000000000017a914d45cb1adffb5215a42720532a076f02c7c778c908738c94d010000000017a914fd237ff37cc91b87b7cef8408ade1bc14a4e67158740787d010000000017a914931fec54c1fea86e574462cc32013f5400b891298780b2e60e0000000017a9147d46a730d31f97b1930d3368a967c309bd4d136a87000000",
      "hash": "f00a46d38fccf8b48554c1f4651eec6bbc294c79ac2e2bc91e63bfcdc574eaa9",
      "authdigest": "8231fdfe4e63846249d0880168deb69bcf7d81cc2f762937677f203fa8d8fd2a",
      "depends": [],
      "fee": 0,
      "sigops": 0,
      "required": true
    },
    "target": "",
    "mintime": 0,
    "mutable": ["time", "transactions", "prevblock"],
    "noncerange": "00000000ffffffff",
    "sigoplimit": 20000,
    "sizelimit": 2000000,
    "curtime": 0,
    "bits": "",
    "height": 0
  },
  "id": 123
}
```

**Comment by arya2** [comment]
@Mergifyio update

**Review by dconnolly — APPROVED** [review]

**Pull request merged**